File: Python/18. 4Sum.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []
        nums.sort()
        combinations = []
        for i in range(0,len(nums)-3,1):
            for j in range(len(nums)-1,i+2,-1):
                left = i+1
                right = j-1
                while right>left:
                    total = nums[i]+nums[j]+nums[left]+nums[right]
                    if total>target:
                        right-=1
                    elif total<target:
                        left+=1
                    else:
                        combination = [nums[i],nums[left],nums[right],nums[j]]
                        if combination not in combinations:
                            combinations.append(combination)
                        right-=1
        return combinations
                            

if __name__ == "__main__":
    nums = [1,0,-1,0,-2,2]
    target = 0
    result = Solution().fourSum(nums,target)
    print(result)

File: CPP/18. 4Sum.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
	vector<vector<int>> fourSum(vector<int> nums, int target) {
		if (nums.size() < 4) {
			return {};
		}
		sort(nums.begin(), nums.end());
		vector<vector<int>> combinations;
		for (size_t i=0; i < nums.size() - 3; i++) {
			if (i > 0 && nums[i] == nums[i - 1]) continue;
			for (size_t j = nums.size()-1; j > i + 2;j--) {
				if (j < nums.size() - 1 && nums[j] == nums[j + 1]) continue;
				int left = i + 1, right = j - 1;
				while (right > left) {
					int total = nums[i] + nums[left] + nums[right] + nums[j];
					if (total > target) {
						right--;
					}
					else if (total < target) {
						left++;
					}
					else {
						vector<int> combination = { nums[i],nums[left],nums[right],nums[j] };
						if (find(combinations.begin(), combinations.end(), combination) == combinations.end()) {
							combinations.push_back(combination);
						}
						right--;
					}
				}
			}
		}
		return combinations;
	}
};

int main() {
	Solution sol;
	vector<int> nums = { 1,0,-1,0,-2,2 };
	int target = 0;
	printVectorOfVectors(sol.fourSum(nums, target));
}